test/tracer.js

```diff
@@ -0,0 +1,183 @@
+import Tracer from '../src/tracer.js';
+import Utils from '../src/util.js';
+
+describe('tracer should', () => {
+    let tracer;
+    let reporter = new InMemoryReporter();
+
+    beforeEach(() => {
+        tracer = new Tracer(
+            'test-service-name',
+            reporter,
+            new ConstSampler(true)
+        );
+    });
+
+    it('create a span correctly through _startInternalSpan', () => {
+        let traceId = Utils.encodeInt64(1);
+        let spanId = Utils.encodeInt64(2);
+        let parentId = Utils.encodeInt64(3);
+        let flags = 1;
+        let context = new SpanContext(traceId, spanId, parentId, flags);
+        let start = Utils.getTimestampMicros();
+        let rpcServer = false;
+        let tags = {
+            'keyOne': 'leela',
+            'keyTwo': 'bender'
+        };
+        let span = tracer._startInternalSpan(context, 'op-name', start, tags, rpcServer);
+
+        assert.isOk(bufferEqual(span.context().traceId, traceId));
+        assert.isOk(bufferEqual(span.context().spanId, spanId));
+        assert.isOk(bufferEqual(span.context().parentId, parentId));
+        assert.equal(span.context().flags, flags);
+        assert.equal(span._start, start);
+        assert.isNotOk(span._firstInProcess);
+        assert.equal(span._binaryAnnotations.length, 2);
+    });
+
+    it ('report a span with tracer level tags', () => {
+        let traceId = Utils.encodeInt64(1);
+        let spanId = Utils.encodeInt64(2);
+        let parentId = Utils.encodeInt64(3);
+        let flags = 1;
+        let startTime = Utils.getTimestampMicros();
+
+        let context = new SpanContext(traceId, spanId, parentId, flags);
+        let span = tracer._startInternalSpan(context, 'op-name', startTime);
+
+        tracer._report(span);
+
+        assert.isOk(reporter.spans.length, 1);
+        assert.equal(span, reporter.spans[0]);
+    });
+
+    it ('start a root span with proper structure', () => {
+        let startTime = new Date(2016, 8, 18).getTime();
+        let span = tracer.startSpan({
+            operationName: 'test-name',
+            startTime: startTime
+        });
+
+        assert.equal(span.context().traceId, span.context().spanId);
+        assert.isNotOk(span.context().parentId);
+        assert.equal(span.context().flags, constants.SAMPLED_MASK);
+        assert.equal(span._start, startTime);
+    });
+
+    it ('start a child span represented as a separate span from parent', () => {
+        let traceId = Utils.encodeInt64(1);
+        let spanId = Utils.encodeInt64(2);
+        let parentId = Utils.encodeInt64(3);
+        let flags = 1;
+        let context = new SpanContext(traceId, spanId, parentId, flags);
+        let startTime = Utils.getTimestampMicros();
+
+        let span = tracer.startSpan({
+            operationName: 'test-name',
+            childOf: context,
+            references: [new opentracing.Reference(opentracing.REFERENCE_CHILD_OF, context)],
```

Comment:
reference is redundant in the presence of `childOf`. Should really be two tests, one with `childOf`, another the same but via reference.